onnxruntime/core/providers/coreml/builders/impl/reduction_op_builder.cc

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Shukant Pal.
+// Licensed under the MIT License.
+
+#include "core/providers/common.h"
+#include "core/providers/shared/utils/utils.h"
+
+#ifdef __APPLE__
+#include "core/providers/coreml/builders/model_builder.h"
+#endif
+#include "core/providers/coreml/builders/helper.h"
+#include "core/providers/coreml/builders/op_builder_factory.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class ReductionOpBuilder : public BaseOpBuilder {
+ private:
+#ifdef __APPLE__
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override;
+#endif
+
+  bool IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                         const logging::Logger& logger) const override;
+};
+
+namespace {
+template <typename T>
+void AddReductionParams(T* params, const std::vector<int64_t>& axes, bool keepdims, bool noop_with_empty_axes) {
+  params->set_keepdims(keepdims);
+
+  for (auto& axis : axes)
+    params->add_axes(axis);
+
+  if (axes.size() == 0 && noop_with_empty_axes)
+    params->reduceall();
+}
+} // namespace
+
+#ifdef __APPLE__
+Status ReductionOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                                                 const logging::Logger& /* logger */) const {
+  const auto& op_type(node.OpType());
+  const auto& input_defs(node.InputDefs());
+  const auto& initializers(model_builder.GetInitializerTensors());
+
+  std::vector<int64_t> axes;
+
+  NodeAttrHelper helper(node);
+  if (input_defs.size() > 1) {
+    auto& axes_tensor = *initializers.at(input_defs[1]->Name());
+    const int64_t* raw_axes = axes_tensor.int64_data().empty()
+      ? reinterpret_cast<const int64_t*>(axes_tensor.raw_data().data())
+      : axes_tensor.int64_data().data();
+    const auto size = axes_tensor.dims()[0];
```

Comment:
Not sure how that'd work. DataAsSpan returns a gsl::span over a std::vector.

Is that the preferred container to use? Since the Initializer is in an if scope, it would be destroyed before the axes data is used so need to keep that in mind as well.

Reply:
sorry, to clarify, was suggesting to use the helper Initializer type instead of dealing with the TensorProto details. you can still copy from the span into the axes vector.

Reply:
generally, gsl::span and InlinedVector use is preferred as appropriate.
https://github.com/microsoft/onnxruntime/blob/9db24f8fec40c047044094441c90bedfe05bae6f/docs/Coding_Conventions_and_Standards.md?plain=1#L15
https://github.com/microsoft/onnxruntime/blob/9db24f8fec40c047044094441c90bedfe05bae6f/docs/Coding_Conventions_and_Standards.md?plain=1#L71

however, when the CoreML EP was initially written those guidelines may not have been in place yet. using std::vector is fine and consistent here.

Reply:
done!